Optimise logs

fpozzobon opened the pull request:
Usage of logs slows down significantly generation of the docs, adding a cache mechanism

fpozzobon committed:
Optimise logs

Usage of logs slows down significantly generation of the docs, adding a cache mechanism

timvink commented on src/mkdocs_git_authors_plugin/git/page.py:
```diff
@@ -238,18 +240,29 @@ def _process_git_log(self, sha, commit) -> None:
+                self._authors.append(co_author)
+            co_author.add_lines(self, commit)
+
+    def _get_git_log(self, sha, commit):
```
Can we rename this function? It's only used for parsing co-authors?

Also, we should have already parsed all the commit earlier: 

https://github.com/timvink/mkdocs-git-authors-plugin/blob/370ee36afc3f0c7a2add374bec73862088cfc419/src/mkdocs_git_authors_plugin/git/repo.py#L116

It's better to extract and add  the co-author information in the Commit class:

https://github.com/timvink/mkdocs-git-authors-plugin/blob/master/src/mkdocs_git_authors_plugin/git/commit.py

then use it here. You won't need to have the caching in that case, and it should be even faster as you're not running `git log` again.

timvink replied:
Yeah I didn't read up enough, my bad. Your approach seems very reasonable. We're running `git blame`, and you added (a cached) `git_log` to get the co-author information. Either way, we're going to have to make a lot more `git` calls, which is a lot slower.

Maybe the better option is to not only have the caching, but also make processing co-authors optional (and disabled by default). The reason being is that the mainstream user flows will take a big performance hit to support a niche feature (git co-authors are not very common). 

Thoughts?

fpozzobon committed:
Code review:
- moving responsibility of deducing co-author to repo
- added co-author inside commit

fpozzobon committed:
list backward compatible

fpozzobon committed:
moving co-author filter to page responsibility

fpozzobon committed:
Add co authors as option

fpozzobon committed:
Adding documentation of co_authors and set to false by default

timvink commented:
Yes, doing it now.

timvink commented:
https://github.com/timvink/mkdocs-git-authors-plugin/releases/tag/v0.10.0
